godot-core/src/builtin/dictionary.rs

```diff
@@ -0,0 +1,314 @@
+    /// it will not.
+    /// 
+    /// This is equivalent to `merge` in godot.
+    pub fn extend_dictionary(&mut self, other: Self, overwrite: bool) {
+        self.as_inner().merge(other, overwrite)
+    }
+
+    /// Returns the number of entries in the dictionary
+    /// 
+    /// This is equivalent to `size` in godot.
+    pub fn len(&self) -> usize {
+        self.as_inner().size().try_into().unwrap()
+    }
+
+    /// Get the pointer corresponding to the given key in the dictionary,
+    /// this pointer is null if there is no value at the given key.
+    pub fn get_ptr<K: ToVariant>(&self, key: K) -> *const Variant {
+        let key = key.to_variant();
+        unsafe {
+            (interface_fn!(dictionary_operator_index_const))(self.sys_const(), key.var_sys_const()) as *const Variant
+        }
+    }
+
+    /// Get the pointer corresponding to the given key in the dictionary,
+    /// if there exists no value at the given key then a new one is created
+    /// and initialized to a nil variant.
+    pub fn get_ptr_mut<K: ToVariant>(&mut self, key: K) -> *mut Variant {
+        let key = key.to_variant();
+        unsafe {
+            let ptr =
+                (interface_fn!(dictionary_operator_index))(self.sys_mut(), key.var_sys_const()) as *mut Variant;
+            // dictionary_operator_index initializes the value so it wont be null
+            assert!(!ptr.is_null());
+            // i think it might be safe to turn this into a &mut Variant?
+            ptr
+        }
+    }
+
+    /// Insert a value at the given key, returning the value
+    /// that previously was at that key if there was one.
+    pub fn insert<K: ToVariant>(&mut self, key: K, value: Variant) -> Option<Variant> {
+        let key = key.to_variant();
+        let old_value = self.remove(key.clone());
+        self.set(key, value);
+        old_value
+    }
+
+    /// Set a key to a given value
+    pub fn set<K: ToVariant>(&mut self, key: K, value: Variant) {
+        let key = key.to_variant();
+        unsafe {
+            *self.get_ptr_mut(key) = value;
+        }
+    }
+
+    /// Convert this dictionary to a strongly typed rust `HashMap`. If the conversion
+    /// fails for any key or value, an error is returned. 
+    pub fn try_to_hashmap<K: FromVariant + Eq + std::hash::Hash, V: FromVariant>(
+        &self,
+    ) -> Result<HashMap<K, V>, VariantConversionError> {
+        let mut map = HashMap::new();
+        for key in self.keys().into_iter() {
+            map.insert(key.try_to()?, self.get(key).unwrap().try_to()?);
+        }
+        Ok(map)
+    }
+
+    /// Convert the keys of this dictionary to a strongly typed rust `HashSet`. If the 
+    /// conversion fails for any key, an error is returned. 
+    pub fn try_to_hashset<K: FromVariant + Eq + std::hash::Hash>(
+        &self,
+    ) -> Result<HashSet<K>, VariantConversionError> {
+        Ok(self.keys().try_to_vec::<K>()?.into_iter().collect())
+    }
```

Comment:
These seem really specific as inherent methods. Perhaps they are better made into impls of the standard [`TryFrom`](https://doc.rust-lang.org/std/convert/trait.TryFrom.html) trait instead?

Also note that you can collect iterators of `Result`s naturally in Rust: `dict.iter().map(T::from_variant).collect::<Result<HashMap<_, _>, _>>()`.

Reply:
Probably this follows the `Array::try_to_vec` API; I agree that `TryFrom` could be more idiomatic though.

@ttencate did you decide explicitly against it for certain reason or was an inherent method more straightforward/expressive?

Reply:
A `TryFrom` impl would be better! Just didn't think of it, I guess. I'll roll that into the `TypedArray` work.

Indeed `HashMap` is maybe a bit too specific here. An iterator over `(key, value)` tuples is more idiomatic, and would allow the user to `collect()` into a `HashMap`, `FxHashMap`, `BTreeMap`, and so on. You'd need a fallible iterator, `Iterator<Item = Result<(K, V), VariantConversionError>>`, which nicely allows to `collect()` into a `Result<HashMap<K, V>, VariantConversionError>` that can subsequently be `unwrap()`ped or otherwise dealt with.

_However_... as we've seen with `Array`, iteration over these shared-reference containers is tricky. With `Array` we could at least make modification during iteration sound, but with `Dictionary` I don't see how you would do that (unless Godot itself somehow detects concurrent modification).

For now, I think the most useful sound approach would be something along these lines:

```
impl<K, V> TryFrom<&Dictionary> for T
where
    T: FromIterator<(K, V)>,
    K: FromVariant,
    V: FromVariant
```

Reply:
I replaced them with `TryFrom` for now, when iteration is added for dictionaries these wont be as necessary i imagine.